cmd/labels_cmd.go

```diff
@@ -0,0 +1,255 @@
+package cmd
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+
+	"github.com/codegangsta/cli"
+	"github.com/ingrammicro/concerto/api/labels"
+	"github.com/ingrammicro/concerto/utils"
+	"github.com/ingrammicro/concerto/utils/format"
+)
+
+// WireUpLabel prepares common resources to send request to Concerto API
+func WireUpLabel(c *cli.Context) (ds *labels.LabelService, f format.Formatter) {
+
+	f = format.GetFormatter()
+
+	config, err := utils.GetConcertoConfig()
+	if err != nil {
+		f.PrintFatal("Couldn't wire up config", err)
+	}
+	hcs, err := utils.NewHTTPConcertoService(config)
+	if err != nil {
+		f.PrintFatal("Couldn't wire up concerto service", err)
+	}
+	ds, err = labels.NewLabelService(hcs)
+	if err != nil {
+		f.PrintFatal("Couldn't wire up label service", err)
+	}
+
+	return ds, f
+}
+
+// LabelList subcommand function
+func LabelList(c *cli.Context) error {
+	debugCmdFuncInfo(c)
+
+	labelsSvc, formatter := WireUpLabel(c)
+	labels, err := labelsSvc.GetLabelList()
+	if err != nil {
+		formatter.PrintFatal("Couldn't receive labels data", err)
+	}
+
+	if err = formatter.PrintList(labels); err != nil {
+		formatter.PrintFatal("Couldn't print/format result", err)
+	}
+	return nil
+}
+
+// LabelCreate subcommand function
+func LabelCreate(c *cli.Context) error {
+	debugCmdFuncInfo(c)
+
+	labelsSvc, formatter := WireUpLabel(c)
+	checkRequiredFlags(c, []string{"name", "resource_type"}, formatter)
+	label, err := labelsSvc.CreateLabel(utils.FlagConvertParams(c))
+	if err != nil {
+		formatter.PrintFatal("Couldn't create label", err)
+	}
+	if err = formatter.PrintItem(*label); err != nil {
+		formatter.PrintFatal("Couldn't print/format result", err)
+	}
+	return nil
+}
+
+// LabelFiltering subcommand function receives an interface representing a collection of labelable resources (Server, Template, ...)
+// Evaluates the matching of assigned labels with the labels requested for filtering.
+func LabelFiltering(c *cli.Context, items interface{}) (*[]interface{}, error) {
```

Comment:
The code for LabelFiltering here and LabelAssignNamesForIDs below works as expected but relays heavily on reflection and exposes too broad an interface using the generic interface{} for input and output types. I think an alternative that uses type composition may be possible and more idiomatic.

Labelable resources would be composed with a LabelableField struct containing the LabelIDs and Labels slices of strings. 
```Go
type LabelableField struct {
  LabelIDs              []string         `json:"label_ids" header:"LABEL_IDS" show:"nolist,noshow"`
  Labels                 []string         `json:"labels" header:"LABELS"`
}

type FirewallProfile struct {
ID string
[...]
LabelableFields
}
```
We would also have a Labelable interface with a set of methods all labelable resources should implement. I'm thinking of a label ID filter and a label IDs and a label names assigning methods but there may be other methods we have to include in it. 
```Go
type Labelable interface {
 FilterByLabelIDs(labelIDs []string) bool
 AssignLabelIDs(labelIDs []string)
 FillInLabelNames(labelNamesByID map[string]string)
}
```
By making LabelableFields implement Labelable, we ensure the different labelable resources (Server, FirewallProfile, etc) implement Labelable:
```Go
func (lf *LabelableFields) FilterByLabelIDs(labelIDs []string) bool {
  for _, lid := range labelIDs {
    var labelIDFound bool
    for _, resourceLabelID := range lf.LabelIDs {
       if lID == resourceLabelID {
         labelIDFound = true
         break;
       }
    }
    if !labelIDFound {
      return false
   }
  }
  return true
}

[...]
```

That way, LabelFiltering could take and return slices of Labelable as input and output. We would still need to convert slices of specific labelable resources (slices of servers, slices of firewall profiles, etc) into slices of labelables, but we would benefit from static typing and clearer code I think...

Reply:
Done